(towards #2966) Add `--enble-cache` psyclone command flag

sergisiso opened the pull request.

sergisiso commented:
@arporter This is ready for a first review.

Note that the integration test of this branch would technically work, it has `RESOLVE_IMPORTS = True` but will be very slow trying to re-parse all dependencies. To make it fast it requires changing a file outside the repo, the `nemo/mk/sct_psyclone.sh` to add a `--enable-cache`, but this will break integration on other branches with a "invalid flag provided". So we need to coordinate when we do this change.

sergisiso commented:
@arporter Fixed (it was a small mistake/omission in the frontend), now is ready for review.

arporter commented on doc/user_guide/psyclone_command.rst:
```diff
@@ -73,6 +73,9 @@ by the command:
                            filename of a PSyclone optimisation recipe
      -I INCLUDE, --include INCLUDE
                            path to Fortran INCLUDE or module files
+     --enable-cache        whether to enable caching imported module dependencies (if
```
"...caching of imported..."

sergisiso replied:
Done

arporter commented on doc/user_guide/psyclone_command.rst:
```diff
@@ -164,6 +166,15 @@ an appropriate error. For example:
+dependencies and the imports happen from multiple files, it can increase the
+psyclone processing time considerably. In this case it is recommended to use
+the `--enable-cache` flag. This will creates a `filename.psycache` file in the
+same location as the original file for every import followed, the next time
```
"...followed. The next ..."

sergisiso replied:
Done

arporter commented on doc/user_guide/psyclone_command.rst:
```diff
@@ -164,6 +166,15 @@ an appropriate error. For example:
+psyclone processing time considerably. In this case it is recommended to use
+the `--enable-cache` flag. This will creates a `filename.psycache` file in the
+same location as the original file for every import followed, the next time
+the same import is requested, if the hash match, the cached file will be used.
```
"hashes"

sergisiso replied:
Done

arporter commented on doc/user_guide/user_scripts.rst:
```diff
@@ -102,9 +104,11 @@ is applied). These are:
-    # creating a PSyIR tree in order to obtain better external symbol information.
-    # It can also be a list of module names for more precise control
+    # Whether to chase the imported modules to improve symbol information (it can
+    # also be a list of module filenames chase only specific modules). This has to
```
"...filenames to limit the chasing to only specific modules"

sergisiso replied:
Done

arporter commented on doc/user_guide/user_scripts.rst:
```diff
@@ -102,9 +104,11 @@ is applied). These are:
-    # It can also be a list of module names for more precise control
+    # Whether to chase the imported modules to improve symbol information (it can
+    # also be a list of module filenames chase only specific modules). This has to
+    # be used in combination of '-I' command flag in order to point to the module
```
"...combination with the '-I'..."

arporter commented on doc/user_guide/user_scripts.rst:
```diff
@@ -102,9 +104,11 @@ is applied). These are:
+    # Whether to chase the imported modules to improve symbol information (it can
+    # also be a list of module filenames chase only specific modules). This has to
+    # be used in combination of '-I' command flag in order to point to the module
+    # location directory and we strongly suggest using it in combination with the
```
"...directory. We also strongly recommend using the '--enable-cache' flag to reduce the performance overhead."

sergisiso replied:
Done

arporter commented on examples/nemo/scripts/acc_kernels_trans.py:
```diff
@@ -77,9 +77,12 @@
-# List of all module names that PSyclone will chase during the creation of the
-# PSyIR tree in order to use the symbol information from those modules
-RESOLVE_IMPORTS = NEMO_MODULES_TO_IMPORT
+# Whether to chase the imported modules to improve symbol information (it can
```
Same suggestions as in `user_scripts.rst`

sergisiso replied:
Done

arporter commented on examples/nemo/scripts/acc_loops_trans.py:
```diff
@@ -39,17 +39,20 @@
+# also be a list of module filenames chase only specific modules). This has to
+# be used in combination of '-I' command flag in order to point to the module
+# location directory and we strongly suggest using it in combination with the
+# '--enable-cache' flag.
```
and again.

sergisiso replied:
Done

arporter commented on examples/nemo/scripts/omp_cpu_trans.py:
```diff
@@ -40,17 +40,19 @@
+# also be a list of module filenames chase only specific modules). This has to
+# be used in combination of '-I' command flag in order to point to the module
+# location directory and we strongly suggest using it in combination with the
+# '--enable-cache' flag.
```
and again.

sergisiso replied:
Done

arporter commented on examples/nemo/scripts/omp_gpu_trans.py:
```diff
@@ -58,9 +58,12 @@
+# also be a list of module filenames chase only specific modules). This has to
+# be used in combination of '-I' command flag in order to point to the module
+# location directory and we strongly suggest using it in combination with the
+# '--enable-cache' flag.
```
ditto

arporter commented on examples/nemo/scripts/utils.py:
```diff
@@ -180,8 +174,8 @@ def _it_should_be(symbol, of_type, instance):
-    information from external modules. Setting NEMO_MODULES_TO_IMPORT above
-    improve the situation but its not complete (not all symbols are imported)
+    information from external modules. Using RESOLVE_IMPORTS improves the
+    situation but its not complete (not all symbols are imported)
```
"it's"

sergisiso replied:
Fixed

arporter commented on src/psyclone/generator.py:
```diff
@@ -437,6 +437,12 @@ def main(arguments):
         help='path to Fortran INCLUDE or module files')
+    parser.add_argument(
+        '--enable-cache', action="store_true", default=False,
+        help='whether to enable caching imported module dependencies (if '
```
Please update as per my comments in the corresponding documentation file.

sergisiso replied:
Done

arporter commented on src/psyclone/generator.py:
```diff
@@ -505,6 +511,11 @@ def main(arguments):
 
+    # This has be be before the Config.get, because otherwise that creates a
+    # ModuleManager Singelton without caching
+    if args.enable_cache:
```
Picky but could just be:
```
_ = ModuleManager.get(cache_active=args.enable_cache)
```

sergisiso replied:
Done

arporter reviewed (changes requested):
Nice Sergi, thanks for this. Pretty much all just tweaks to the comments :-)

arporter commented:
I should also have asked: how does the processing time for NEMO with this PR compare to what it was before?

sergisiso commented on doc/user_guide/psyclone_command.rst:
```diff
@@ -73,6 +73,9 @@ by the command:
                            path to Fortran INCLUDE or module files
+     --enable-cache        whether to enable caching imported module dependencies (if
+                           enabled, it will generate a .psycache file of each imported
+                           module in the same location as the import source file).
```
Done

sergisiso commented on src/psyclone/generator.py:
```diff
@@ -505,6 +511,11 @@ def main(arguments):
             sys.exit(1)
 
+    # This has be be before the Config.get, because otherwise that creates a
+    # ModuleManager Singelton without caching
```
Fixed

sergisiso commented:
@arporter This is ready for another review, I reverted the NEMO "resolve all imports" and converted this PR to a towards, with the current limitations listed in the issue. But I will still want to continue with this to get as I will increase the number of imports to attempt the name-parenthesis defaulting to Call unless resolved.

arporter commented:
Integration tests were fine.
